CARDS-1830: Creating forms via API ignores the requiredSubjectTypes setting

sofi2002sofi opened the pull request:
[CARDS-1830](https://phenotips.atlassian.net/browse/CARDS-1830)

sofi2002sofi committed:
CARDS-1830: Creating forms via API ignores the requiredSubjectTypes setting

sofi2002sofi committed:
CARDS-1830: Creating forms via API ignores the requiredSubjectTypes setting

remove dublicated array in feature.json

IntegralProgrammer commented on modules/data-entry/src/main/java/io/uhndata/cards/internal/RequiredSubjectTypesValidator.java:
```diff
@@ -0,0 +1,138 @@
+ */
+package io.uhndata.cards.internal;
+
+import java.util.*;
```
Please only import the members of the `java.util` package that are used within this file instead of all members from the `java.util` package. This actually causes the build to fail with an `AvoidStarImport` checkstyle violation.

IntegralProgrammer commented on modules/data-entry/src/main/java/io/uhndata/cards/internal/RequiredSubjectTypesValidator.java:
```diff
@@ -0,0 +1,138 @@
+        final String subjectUUID = after.getProperty("subject").getValue(Type.REFERENCE).toString();
+        LOGGER.warn("Added this --> {}", after);
+        LOGGER.warn("A cards:Form node was just added with questionnaire={} and subject={} !!!", questionnaireUUID,
+                subjectUUID);
```
Debugging statements used in development should not be added into the `dev` branch. Please remove these.

IntegralProgrammer commented on modules/data-entry/src/main/java/io/uhndata/cards/internal/RequiredSubjectTypesValidator.java:
```diff
@@ -0,0 +1,138 @@
+
+            // Get the Subject Resource associated with this Form
+            final Resource subject = getSubjectResourceByUuid(serviceResolver, subjectUUID);
+            LOGGER.warn("Resolved /Subjects/{} --> {}", subjectUUID, subject);
```
Debugging statements used in development should not be added into the `dev` branch. Please remove.

IntegralProgrammer commented on modules/data-entry/src/main/java/io/uhndata/cards/internal/RequiredSubjectTypesValidator.java:
```diff
@@ -0,0 +1,138 @@
+
+            // Get the type value for the Subject
+            final String type = subject.getValueMap().get("type", "null");
+            LOGGER.warn("Questionnaire type --> {}", type);
```
Debugging statements used in development should not be added into the `dev` branch. Please remove.

IntegralProgrammer commented on modules/data-entry/src/main/java/io/uhndata/cards/internal/RequiredSubjectTypesValidator.java:
```diff
@@ -0,0 +1,138 @@
+            LOGGER.warn("Questionnaire type --> {}", type);
+            // Get the Questionnaire Resource associated with this Form
+            final Resource questionnaire = getQuestionnaireResourceByUuid(serviceResolver, questionnaireUUID);
+            LOGGER.warn("Resolved /Questionnaires/{} --> {}", questionnaireUUID, questionnaire);
```
Debugging statements used in development should not be added into the `dev` branch. Please remove.

IntegralProgrammer commented on modules/data-entry/src/main/java/io/uhndata/cards/internal/RequiredSubjectTypesValidator.java:
```diff
@@ -0,0 +1,138 @@
+            // Get the requiredSubjectTypes value for the Questionnaire
+            final String requiredSubjectTypes = questionnaire.getValueMap().get("requiredSubjectTypes", "null");
+            final List<String> allRequiredSubjectTypes = Arrays.asList(requiredSubjectTypes.split(","));
+            LOGGER.warn("Listened by the Questionnaire types --> {}", allRequiredSubjectTypes);
```
Debugging statements used in development should not be added into the `dev` branch. Please remove.

IntegralProgrammer commented on modules/data-entry/src/main/java/io/uhndata/cards/internal/RequiredSubjectTypesValidator.java:
```diff
@@ -0,0 +1,138 @@
+    }
+
+    @Override
+    public @Nullable Validator childNodeAdded(String s, NodeState after) throws CommitFailedException
```
```suggestion
    public @Nullable Validator childNodeAdded(String name, NodeState after) throws CommitFailedException
```

IntegralProgrammer commented on modules/data-entry/src/main/java/io/uhndata/cards/internal/RequiredSubjectTypesValidator.java:
```diff
@@ -0,0 +1,138 @@
+            }
+
+            // Get the requiredSubjectTypes value for the Questionnaire
+            final String requiredSubjectTypes = questionnaire.getValueMap().get("requiredSubjectTypes", "null");
+            final List<String> allRequiredSubjectTypes = Arrays.asList(requiredSubjectTypes.split(","));
```
The `requiredSubjectTypes` property is multi-valued and should be obtained from the JCR accordingly. See https://github.com/data-team-uhn/cards/blob/fd8b362bf792ed54d9a31e4b35429ef85a3b44f7/proms-resources/backend/src/main/java/io/uhndata/cards/proms/emailnotifications/AppointmentUtils.java#L83 as an example of a better way to do this.

IntegralProgrammer commented on modules/data-entry/src/main/java/io/uhndata/cards/internal/RequiredSubjectTypesValidator.java:
```diff
@@ -0,0 +1,138 @@
+            }
+
+            // Get the type value for the Subject
+            final String type = subject.getValueMap().get("type", "null");
```
```suggestion
            final String type = subject.getValueMap().get("type", "");
```
would be more consistent with how this is done in other places in the project.

IntegralProgrammer reviewed (changes requested):
After fixing the `AvoidStarImport` checkstyle issue, I was able to build and test. The code's functionality was correct but the style could be improved as further detailed by my comments.

sofi2002sofi committed:
CARDS-1830: Creating forms via API ignores the requiredSubjectTypes setting

remove debbuging statements, fix java.util imports, edit the method of assigning properties value to the array

IntegralProgrammer commented on modules/data-entry/src/main/java/io/uhndata/cards/internal/RequiredSubjectTypesValidator.java:
```diff
@@ -0,0 +1,131 @@
+
+            // Get the requiredSubjectTypes value for the Questionnaire
+            final List<String> allRequiredSubjectTypes = questionnaire.getValueMap().get("requiredSubjectTypes",
+                    List.of());
```
This will always return an empty `List<String>` and therefore will always cause the creation of a new Form to fail. You would want to do something like

```suggestion
            final List<String> allRequiredSubjectTypes = Arrays.asList(
                questionnaire.getValueMap().get("requiredSubjectTypes", String[].class)
            );
```

instead - be sure to `import java.util.Arrays`.

IntegralProgrammer commented on modules/data-entry/src/main/java/io/uhndata/cards/internal/RequiredSubjectTypesValidator.java:
```diff
@@ -0,0 +1,131 @@
+import org.apache.sling.api.resource.Resource;
+import org.apache.sling.api.resource.ResourceResolver;
+import org.apache.sling.api.resource.ResourceResolverFactory;
+import org.jetbrains.annotations.Nullable;
```
This does not need to be imported.

IntegralProgrammer commented on modules/data-entry/src/main/java/io/uhndata/cards/internal/RequiredSubjectTypesValidator.java:
```diff
@@ -0,0 +1,131 @@
+    }
+
+    @Override
+    public @Nullable Validator childNodeAdded(String name, NodeState after) throws CommitFailedException
```
```suggestion
    public Validator childNodeAdded(String name, NodeState after) throws CommitFailedException
```

IntegralProgrammer commented on modules/data-entry/src/main/java/io/uhndata/cards/internal/RequiredSubjectTypesValidator.java:
```diff
@@ -0,0 +1,131 @@
+    }
+
+    @Override
+    public @Nullable Validator childNodeChanged(String s, NodeState before, NodeState after)
```
```suggestion
    public Validator childNodeChanged(String s, NodeState before, NodeState after)
```

IntegralProgrammer reviewed (dismissed):
Code does not work as is - it always fails to create a new Form as `List<String> allRequiredSubjectTypes` is always empty, but after the suggestions are applied, all should be good to go.

IntegralProgrammer commented on modules/data-entry/src/main/java/io/uhndata/cards/internal/RequiredSubjectTypesValidator.java:
```diff
@@ -0,0 +1,131 @@
+        return this;
+    }
+
+    /**
+     * Obtains the Questionnaire Resource that has a specified jcr:uuid.
+     *
+     * @param ResourceResolver a ResourceResolver that can be used for querying the JCR
+     * @param uuid the jcr:uuid of the Questionnaire Resource which we wish to obtain
+     * @return the matching Questionnaire Resource or null if none can be found
+     */
+    private Resource getQuestionnaireResourceByUuid(final ResourceResolver serviceResolver, final String uuid)
+    {
+        final Iterator<Resource> resourceIterator = serviceResolver.findResources(
+                "SELECT * FROM [cards:Questionnaire] AS q WHERE q.'jcr:uuid'='" + uuid + "'", "JCR-SQL2");
+
+        if (!resourceIterator.hasNext()) {
+            return null;
+        }
+        return resourceIterator.next();
+    }
+
+    /**
+     * Obtains the Subject Resource that has a specified jcr:uuid.
+     *
+     * @param ResourceResolver a ResourceResolver that can be used for querying the JCR
+     * @param uuid the jcr:uuid of the Subject Resource which we wish to obtain
+     * @return the matching Subject Resource or null if none can be found
+     */
+    private Resource getSubjectResourceByUuid(final ResourceResolver serviceResolver, final String uuid)
+    {
+        final Iterator<Resource> resourceIterator = serviceResolver.findResources(
+                "SELECT * FROM [cards:Subject] as s WHERE s.'jcr:uuid'='" + uuid + "'", "JCR-SQL2");
+        if (!resourceIterator.hasNext()) {
+            return null;
+        }
+        return resourceIterator.next();
+    }
```
Although it's not necessary to make this change in this PR, these two methods are almost identical and therefore the DRY (_Don't Repeat Yourself_) principle could be applied here by creating a new method `getResourceByUuid(final ResourceResolver, serviceResolver, final String resourceType, final String uuid)`.

Furthermore, a _private_ `getQuestionnaireResourceByUuid` is already defined in https://github.com/data-team-uhn/cards/blob/fcffa61fba18adf7098e9ec0217d19b486cd71f8/modules/data-entry/src/main/java/io/uhndata/cards/internal/MaxFormsOfTypePerSubjectValidator.java#L120-L136

Perhaps we should create some kind of helper utility class with this method that can be used by both `MaxFormsOfTypePerSubjectValidator`, `RequiredSubjectTypesValidator`, and any others in the future that may need this. @sdumitriu?

sdumitriu replied:
Isn't `io.uhndata.cards.forms.api.QuestionnaireUtils.getQuestionnaire(String)` exactly that?

Ah, almost, it returns a Node, not a Resource. But could a Node be used here instead of a Resource?

sofi2002sofi committed:
CARDS-1830: Creating forms via API ignores the requiredSubjectTypes setting

fix according to comments

sofi2002sofi committed:
CARDS-1830: Creating forms via API ignores the requiredSubjectTypes setting

created a new method that replaces 2 logically similar methods

sdumitriu commented on modules/data-entry/src/main/java/io/uhndata/cards/internal/RequiredSubjectTypesValidator.java:
```diff
@@ -0,0 +1,117 @@
+                    String[].class);
+            final List<String> allRequiredSubjectTypesList = Arrays.asList(allRequiredSubjectTypes != null
+                    ? allRequiredSubjectTypes : new String[0]);
+            if (!allRequiredSubjectTypesList.contains(type)) {
```
It is possible for a questionnaire not to specify any required subject type, which means any subject type is valid. The condition should be `if (allRequiredSubjectTypesList.size() > 0 && !allRequiredSubjectTypesList.contains(type))`.

sdumitriu commented on modules/data-entry/src/main/java/io/uhndata/cards/internal/RequiredSubjectTypesValidator.java:
```diff
@@ -0,0 +1,117 @@
+    }
+
+    @Override
+    public Validator childNodeAdded(String name, NodeState after) throws CommitFailedException
```
One scenario not taken into account is changing the subject after the form is created. We should also handle that case in a `propertyChanged {if node is form and property is subject}` method.

sdumitriu committed:
CARDS-1830: Creating forms via API ignores the requiredSubjectTypes setting

sdumitriu committed:
[misc] Various build failure fixes

sdumitriu commented:
One thing still not perfect: when trying to create extra forms through the UI, it fails (correctly) but the message is not the one displayed when the UI knows there are enough forms already, but a generic, cryptic `409` error.

marta- commented:
> One thing still not perfect: when trying to create extra forms through the UI, it fails (correctly) but the message is not the one displayed when the UI knows there are enough forms already, but a generic, cryptic `409` error.

This is not about there being enough forms already, but about assigning a form to the wrong subject. Regardless, they both generate the same error message `New form request failed with error code 409: Conflict` (although under some circumstances the latter just fails silently), so we can't really differentiate between them on the UI side to display an informative message. Is it possible to produce different responses on the backend?

marta- commented:
> One thing still not perfect: when trying to create extra forms through the UI, it fails (correctly) but the message is not the one displayed when the UI knows there are enough forms already, but a generic, cryptic `409` error.

This is not about there being enough forms already, but about assigning a form to the wrong subject. Regardless, they both generate the same error message `New form request failed with error code 409: Conflict` (although under some circumstances the latter just fails silently), so we can't really differentiate between them on the UI side to display an informative message. Is it possible to produce different responses on the backend? @sdumitriu
